docs/src/appendix/mast.md

```diff
@@ -0,0 +1,526 @@
+
+##### Basic Blocks
+
+Unlike the other node types, `BasicBlockNode` has a much more complex representation. It must encode the operations contained in the block, as well as decorators for those operations. Both operations and decorators may have associated data of their own. Additionally, operations of a block are batched and grouped to facilitate computation of their Merkle digests.
+
+We do not represent batching/grouping in the binary format, at least in this version. Instead, we specify that operations decoded from the binary format will be batched and grouped using fixed sizes. Specifically, a batch size of 8, and a group size of 9. This could alternatively be handled by storing these sizes in the unused padding bytes in `MastNodeType` which are available to the `Block` variant. For this draft though, I'm making this a property of the specific version of the binary format, rather than per-block configurable.
+
+```rust
+/// Represents a variable-width index into the `strings` table
+///
+/// If the strings table is small, indexes into it are only a
+/// single byte, but if the table grows too large, only those
+/// later strings need pay the price of the larger index.
+pub type StringIndex = VInt64;
+
+/// An entry in the `strings` table of `EncodedMastForest`
+///
+/// We could also use null-terminated strings, but this
+/// allows us to be more precise, and permit any content
+/// in the strings, even NUL bytes.
+#[repr(C)]
+pub struct StringRef {
+    offset: DataOffset,
+    // Maximum allowed length of a string is 2^32
+    len: u32,
+}
+
+/// Unlike most other type definitions here, this one
+/// is not encoded using the usual Rust layout rules.
+///
+/// Specifically, we treat this as a variable-width type,
+/// without the padding that would normally be introduced
+/// to make smaller variants as large as the largest variant.
+///
+/// So each encoded operation starts with the tag byte, which
+/// indicates which variant we're going to be parsing. We then
+/// parse the data for that variant, and immediately look for
+/// the next tag byte (if there are more ops to parse).
+///
+/// It is strictly for purposes of readability.
+///
+/// NOTE: Decorators are always associated with the next non-decorator
+/// operation in the basic block being decoded.
+#[repr(u8)]
+pub enum EncodedOperation {
+    /// A simple operation, with its opcode encoded in a single byte
+    Op { opcode: u8 },
+    /// An operation with an immediate argument.
+    ///
+    /// Like `Op`, the opcode takes a single byte, and the immediate
+    /// is encoded using a variable-width integer. The immediate must
+    /// be converted to/from `Felt` when decoded.
+    OpWithImmediate {
+        opcode: u8,
+        imm: VInt64,
+    },
```

Comment:
I wonder if this distinction is actually needed. By reading the opcode we could tell whether this operation has an immediate or not. Currently, the only operation which has an immediate is the `PUSH` operation.